### ISSUE TITLE
Implement Leases

### ISSUE THREAD

**Issue by shvachko** [GiraffaFS/giraffa#31]
Original [issue 31](https://code.google.com/p/giraffa/issues/detail?id=31) created by shvachko on 2012-09-18T02:18:21.000Z:

We need to implement leases for writing to and completing files. This is a major feature in Hadoop and it needs to be made part of Giraffa.

**Comment by shvachko** [comment]
Comment [#2](https://code.google.com/p/giraffa/issues/detail?id=31#c2) originally posted by shvachko on 2012-09-25T08:08:09.000Z:

Leases should belong to NamespaceProcessor.
In HDFS leases are per client, that is a call to renewLease() from a specific client renews all leases owned by that client. 
Giraffa should do the same. The client will periodically send renewLease() to all RegionServers, which will first hit respective NamespaceProcessors, which should renew all leases owned by the client.

**Comment by shvachko** [comment]
Comment [#3](https://code.google.com/p/giraffa/issues/detail?id=31#c3) originally posted by shvachko on 2012-10-18T18:06:38.000Z:

Bumping the priority up. Which means it is the next thing to do AFTER the current work in progress is done, not that we should drop everything and do that.

**Comment by shvachko** [comment]
Comment [#4](https://code.google.com/p/giraffa/issues/detail?id=31#c4) originally posted by shvachko on 2012-10-23T07:52:01.000Z:

Are leases stored in a data structure? A HashMap I take it.
Are leases a Hadoop object? We will inherit them from Hadoop, correct?

**Comment by shvachko** [comment]
Comment [#5](https://code.google.com/p/giraffa/issues/detail?id=31#c5) originally posted by shvachko on 2012-10-24T04:55:02.000Z:

There is LeaseManager in HDFS. But it is dependent on FSNamesystem. So we will not likely be able to reuse it.
The idea is that each client periodically sends a renewLease() request, which is intended to prolong leases for ALL files the client currently opened for write, those that were not completed. In Giraffa renewLease() request goes to all NamespaceProcessors. Every NamespaceProcessor keeps track of opened files per client. This is called Lease in HDFS.
The following functions should be performed by the LeaseManager.
- Define Lease expiration time (soft and hard).
- Prohibit other clients from writing into files that already have a Lease. Unless it expired.
- Lease monitoring: detecting expired leases.
- Do lease recovery, which happens when lease expired (hard limit) and nobody claimed it.

In Giraffa we also want leases to be persistent, since if a RegionServer dies leases should be transfered to another RS along with the region containing the file.
I think we should have some lease related fields in the file row itself.
So it is pretty big project. Needs a design.

**Comment by shvachko** [comment]
Comment [#6](https://code.google.com/p/giraffa/issues/detail?id=31#c6) originally posted by shvachko on 2013-08-02T20:02:08.000Z:

Should we start with opening a separate issue to create some persisted Lease field in our INodes?

**Comment by shvachko** [comment]
Comment [#7](https://code.google.com/p/giraffa/issues/detail?id=31#c7) originally posted by shvachko on 2014-06-17T15:19:10.000Z:

In HDFS lease recovery, NameNode adds the last block of the file to a recovery queue for the primary DataNode of that file, and this queue is sent to the DataNode during a HeartBeat command. The DataNode recovers the block and uses the DatanodeProtocol (method "commitBlockSynchronization()") to tell the NameNode it is done recovering the block, at which point the file is closed and the lease is removed. Since we do not currently implement the DatanodeProtocol, we have no way of determining when or if the block has been successfully recovered / the file should be closed and the lease should be removed. This issue is blocked until we have BlockManagers that do not rely on NameNode to communicate with DataNodes.

**Comment by shvachko** [comment]
Comment [#8](https://code.google.com/p/giraffa/issues/detail?id=31#c8) originally posted by shvachko on 2014-10-18T03:01:41.000Z:

Lease management will be distributed here. We can rely on INodes to maintain a lease field and update it / query it accordingly. However, it will be difficult to implement an entity whom can definitively say 'this INode's lease is now expired' as multiple NamespaceProcessors might try to contend for the lease.

One thing we can do is to resolve multiple instances of lease recovery would be to make it require a ZK lock. That would ensure that multiple NamespaceProcessors or GiraffaClients contending for lease recovery wouldn't kill the system. 

Lease renewal is pretty straightforward as it is simply an update to the INode field.

I am unsure how to keep track of "opened files" by clients per NamespaceProcessor. We may have to have a background scanner that goes through the INodes (rows) hosted on that RegionServer and check lease of every file. (Perhaps this can also be the Rename clean-up thread?)

**Comment by shvachko** [comment]
Comment [#9](https://code.google.com/p/giraffa/issues/detail?id=31#c9) originally posted by shvachko on 2014-11-12T03:22:10.000Z:

I spent some time looking into this today. I noticed something we need to discuss / address in order to do this work.

The method signature we need to override is pre/post-openRegion(). This method signature only exists in the RegionObserver coprocessor interface.

This implies we need to either convert NamespaceProcessor into a RegionObserver coprocessor or we need to discuss how to implement Lease management into the BlockManagementAgent.

**Comment by shvachko** [comment]
Comment [#10](https://code.google.com/p/giraffa/issues/detail?id=31#c10) originally posted by shvachko on 2014-11-13T05:09:32.000Z:

I have come up with an initial patch and unit test to demonstrate adding Leases to INodes as a FileField at the BlockManagementAgent level.

I also demonstrate making use of an HDFS LeaseManager as the means for keeping track of the leases on the running system.

There is much to discuss -- I will write up a formal design document for this work.

**Comment by shvachko** [comment]
Comment [#11](https://code.google.com/p/giraffa/issues/detail?id=31#c11) originally posted by shvachko on 2014-11-15T05:40:14.000Z:

I discovered that there is a way to access "shared data" between Coprocessors.
In the start() of Coprocessors there is a RegionCoprocessorEnvironment whom one can call getSharedData() on, which returns a ConcurrentMap<String, Object>. 
We can make use of this to pass a "GiraffaState" (as Konst suggested) between the Coprocessors on any particular host.

**Comment by shvachko** [comment]
Comment [#12](https://code.google.com/p/giraffa/issues/detail?id=31#c12) originally posted by shvachko on 2014-11-16T07:12:27.000Z:

My pride and joy. Please take a look at the latest patch. Please excuse the file size, it's mostly because we are introducing new classes (GiraffaState).

**Comment by shvachko** [comment]
Comment [#13](https://code.google.com/p/giraffa/issues/detail?id=31#c13) originally posted by shvachko on 2015-01-08T02:19:50.000Z:

Quick note about RegionScanners. In BlockManagementAgent, RegionScanners should be used in the follow fashion if you're calling getRaw:

RegionScanner scanner = ...
- MultiVersionConsistencyControl.setThreadReadPoint(scanner.getMvccReadPoint());
- region.startRegionOperation();
- try {
-   synchronized(scanner) {
-     ...
-     boolean moreRows = scanner.nextRaw(values);
-     ...
-   }
- } finally {
-   region.closeRegionOperation();
- }
  (taken from the RegionScanner.getRaw comment)
  Will continue reviewing tomorrow

**Comment by shvachko** [comment]
Comment [#14](https://code.google.com/p/giraffa/issues/detail?id=31#c14) originally posted by shvachko on 2015-04-09T22:29:12.000Z:
1. [retracted]
2. NamespaceProcessor.start and BlockManagementAgent.start: all we need to do to load/set the GiraffaState is:
   {{this.giraffaState = (GiraffaState) env.getSharedData().putIfAbsent(GIRAFFA_STATE_SHARED_DATA, new GiraffaState());}}
   3: NamespaceProcessor.create: need to incorporate create flag into logic.
   4: NamespaceProcessor.takeOverlease: Call file.getLease() just once and check if its null only once. Use else-if for second inner if clause.
   5: NamespaceProcessor.complete: Can we use the clientName given as the parameter instead of obtaining it from the login user? If not, we can move the logic for obtaining the clientName into checkLease().
   6: See RegionScanner comment above.
   7: Patch complains about whitespace errors.

**Comment by shvachko** [comment]
Plamen, reviewed your patch. Sounds like the right approach.
1. My main concern is about dependencies on HDFS LeaseManager. I think we should have our own LeaseManager. It can be implemented as a wrapper of hdfs.LeaseManager for now. But in the end we will replace it internally, as generally we cannot rely on private Hadoop interfaces.
2. There is no renewLease() and no lease recovery at the moment, right? So we should further address it.
3. Giraffa LeaseManager should be independent of the underlying namespace impl (currently HBase table). So, for example, resolution of all INodes belonging to the client requesting to renew should be abstract, and implemented as a scan on local regions for HBase.

**Comment by shvachko** [comment]
I refactored Plamen's patch. The main focus was on 1 and 3 from above.
- I renamed GiraffaState to LeaseManager, and moved it under giraffa package. Its implementation uses HLMAdapter, so that hdfs.LeaseManager does not leak outside of it. We can put the real implementation of giraffa.LeaseManager later get rid of HLMAdapter.
- GLease is renamed to FileLease. Moved it to giraffa package as well.

I think both LeaseManager and FileLease don't have any specifics from HDFS or HBase now.

**Comment by shvachko** [comment]
The sharing of LeaseManager between NamespaceProcessor and BlockManagementAgent does not work. This is because sharedData in coprocessors is shared between instances of the same coprocessors, but not between instances of different coprocessors.
Do we really need to share LeaseManager. BlockManagementAgent uses it in prePut() in the patch to remove the lease if the file is closed or deleted. It also updated the lease in case of other put actions.
But. Both close and delete are explicit RPC calls, so we can delete the lease in NamespaceProcessor when executing these calls. And the periodic renewLease() should take care of updating the lease.
We should keep the sharing logic for NamespaceProcessor though. There are multiple NamespaceProcessors created for each table. We only need one NamespaceProcessor for table Namespace, but until we get that, the LeaseManager gotta be shared.

**Comment by shvachko** [comment]
On Plamen's latest patch. I think we are on the right track. Minor things:
1. Looks like you still persist file patch into the lease in the row. I thought we don't need it.
2. Remove commented out lines in NamespaceAgent. Also `NamespaceAgent.renewLease()` needs reformatting.
3. Let's try to avoid white space changes with serious patches.
4. In `NSP.create()` could you add a check that `create` flag is true.
5. Unprivate `checkLease()` and `takeOverLease()`
6. `takeOverLease()` should be `reassignLease()`
7. `HLMAdapter.getLease()` should be `HLMAdapter.getLeases()` - plural.
8. `.NamespaceProcessor.renewLease()` should check that INode is not null.

**Comment by pjeli** [comment]
Addressed all nits but point 1. We don't need it but I am unable to regenerate the Protobuf code at the moment; unsure why. Code is view-able as a pull request.

**Comment by milandesai** [comment]
A few more things:
1. In NamespaceProcessor.start(), we can call originateSharedLeaseManager() rather than NamespaceProcessor.originateSharedLeaseManager().
2. In NamespaceProcesor.originateSharedLeaseManager, we don't need the log parameter as we can access LOG statically. Also remove the parameter from the javadoc.
3. In NamespaceProcessor.originateSharedLeaseManager, if the return value of putIfAbsent() is non-null, we should return that value to avoid a race condition.
4. In NamespaceProcessor.reassignLease, remove the two checks for "lease != null" as they will always return true. Additionally, we don't need to call file.getLease() twice.

**Comment by shvachko** [comment]
>  if the return value of putIfAbsent() is non-null, we should return that value to avoid a race condition.

This does not avoid a race condition. We understand that, right? But we can return.

**Comment by milandesai** [comment]
I meant that its possible another coprocessor has already created the shared lease manager at this point, and if so, putIfAbsent() will return **that** lease manager, which is the one we should be using. Otherwise two coprocessors could be using different LeaseManagers.

**Comment by shvachko** [comment]
`TestGiraffaConcurrentClients` is failing with this patch. The first of the two `testFiveCreation` would succeed, but the second then fails. Did some debugging. I see that on the second  `testFiveCreation` all 5 threads creating files fail with exception:

```
15/05/04 11:52:29 INFO hbase.NamespaceAgent: java.io.IOException: Number of block infos is 0 for /user/shv/zero/8851930161767815205
    at org.apache.giraffa.hbase.NamespaceProcessor.addBlock(NamespaceProcessor.java:228)
    at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolServerSideTranslatorPB.addBlock(ClientNamenodeProtocolServerSideTranslatorPB.java:351)
    at org.apache.giraffa.hbase.ClientNamenodeProtocolServerSideCallbackTranslatorPB.addBlock(ClientNamenodeProtocolServerSideCallbackTranslatorPB.java:237)
    at org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos$ClientNamenodeProtocol$1.addBlock(ClientNamenodeProtocolProtos.java:40406)
    at org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos$ClientNamenodeProtocol.callMethod(ClientNamenodeProtocolProtos.java:41332)
    at org.apache.hadoop.hbase.regionserver.HRegion.execService(HRegion.java:5120)
    at org.apache.hadoop.hbase.regionserver.HRegionServer.execService(HRegionServer.java:3220)
    at org.apache.hadoop.hbase.protobuf.generated.ClientProtos$ClientService$2.callBlockingMethod(ClientProtos.java:20807)
    at org.apache.hadoop.hbase.ipc.RpcServer.call(RpcServer.java:2156)
    at org.apache.hadoop.hbase.ipc.RpcServer$Handler.run(RpcServer.java:1861)
```

In the debugger I can see that the block is created by `BlockManagementAgent`, but it cannot be seen in the INode in `NamespaceProcessor`.

**Comment by pjeli** [comment]
Figured it out after much debate.

Turns out that renewLease was overriding the BlockLocations column since we update ALL the fields at once whenever we call updateINode().
Solution is to do more fine-grained row updates; for now we added an updateINodeLease() method, but in the future we should keep track of only the fields of the INode which need updating..

**Comment by shvachko** [comment]
Let's move the `LOG.info()` message from `prePut()` to the end of `allocateBlock()`. And then remove all other LOG messages from  `allocateBlock()`. They seem to be uninformative or misplaced.
Looks good otherwise.

**Comment by shvachko** [comment]
+1 on the latest patch

**Comment by pjeli** [comment]
Committed. Hash: 4ea994482bcd014ff6321d194b8698cac6c4abfd.
